exchanges/bithumb/bithumb_websocket.go

```diff
@@ -0,0 +1,213 @@
+		lu, err = time.ParseInLocation(tickerTimeLayout,
+			tick.Date+tick.Time,
+			location)
+		if err != nil {
+			return err
+		}
+		b.Websocket.DataHandler <- &ticker.Price{
+			ExchangeName: b.Name,
+			AssetType:    asset.Spot,
+			Last:         tick.PreviousClosePrice,
+			Pair:         tick.Symbol,
+			Open:         tick.OpenPrice,
+			Close:        tick.ClosePrice,
+			Low:          tick.LowPrice,
+			High:         tick.HighPrice,
+			QuoteVolume:  tick.Value,
+			Volume:       tick.Volume,
+			LastUpdated:  lu,
+		}
+	case "transaction":
+		if !b.IsSaveTradeDataEnabled() {
+			return nil
+		}
+
+		var trades WsTransactions
+		err = json.Unmarshal(resp.Content, &trades)
+		if err != nil {
+			return err
+		}
+
+		var lu time.Time
+		for x := range trades.List {
+			lu, err = time.ParseInLocation(tradeTimeLayout,
+				trades.List[x].ContractTime,
+				location)
+			if err != nil {
+				return err
+			}
+
+			err = b.AddTradesToBuffer(trade.Data{
+				Exchange:     b.Name,
+				AssetType:    asset.Spot,
+				CurrencyPair: trades.List[x].Symbol,
+				Timestamp:    lu,
+				Price:        trades.List[x].ContractPrice,
+				Amount:       trades.List[x].ContractAmount,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	case "orderbookdepth":
+		var orderbooks WsOrderbooks
+		err = json.Unmarshal(resp.Content, &orderbooks)
+		if err != nil {
+			return err
+		}
+		init, err := b.UpdateLocalBuffer(&orderbooks)
+		if err != nil && !init {
+			return fmt.Errorf("%v - UpdateLocalCache error: %s", b.Name, err)
+		}
+		return nil
+	default:
+		return fmt.Errorf("unhandled response type %s", resp.Type)
+	}
+
+	return nil
+}
+
+// GenerateSubscriptions generates the default subscription set
+func (b *Bithumb) GenerateSubscriptions() ([]stream.ChannelSubscription, error) {
+	var channels = []string{"ticker", "transaction", "orderbookdepth"}
+	var subscriptions []stream.ChannelSubscription
+	pairs, err := b.GetEnabledPairs(asset.Spot)
+	if err != nil {
+		return nil, err
+	}
+
+	for x := range pairs {
+		for y := range channels {
+			subscriptions = append(subscriptions, stream.ChannelSubscription{
+				Channel:  channels[y],
+				Currency: pairs[x].Format("_", true),
+				Asset:    asset.Spot,
+			})
+		}
+	}
+	return subscriptions, nil
+}
+
+// Subscribe subscribes to a set of channels
+func (b *Bithumb) Subscribe(channelsToSubscribe []stream.ChannelSubscription) error {
+	subs := make(map[string]*WsSubscribe)
+	for i := range channelsToSubscribe {
+		s, ok := subs[channelsToSubscribe[i].Channel]
+		if !ok {
+			s = &WsSubscribe{
+				Type: channelsToSubscribe[i].Channel,
+			}
+			subs[channelsToSubscribe[i].Channel] = s
+		}
+		s.Symbols = append(s.Symbols, channelsToSubscribe[i].Currency)
+	}
+
+	tSub, ok := subs["ticker"]
+	if ok {
+		tSub.TickTypes = wsDefaultTickTypes
+	}
+
+	for _, s := range subs {
+		err := b.Websocket.Conn.SendJSONMessage(s)
+		if err != nil {
+			return err
+		}
+		time.Sleep(time.Second) // Undocumented rate limiting.
```

Comment:
Websockets have rate limiting functionality. Under `return b.Websocket.SetupNewConnection(stream.ConnectionSetup{` in `bithumb,_wrapper.go:178` add the rate limit property